exonum/src/events/noise/mod.rs

```diff
@@ -40,80 +41,109 @@ pub struct HandshakeParams {
     pub max_message_len: u32,
 }
 
+pub trait Handshake {
+    fn listen<S: AsyncRead + AsyncWrite + 'static>(self, stream: S) -> HandshakeResult<S>;
+    fn send<S: AsyncRead + AsyncWrite + 'static>(self, stream: S) -> HandshakeResult<S>;
+}
+
 #[derive(Debug)]
-pub struct NoiseHandshake {}
+pub struct NoiseHandshake {
+    noise: NoiseWrapper,
+    max_message_len: u32,
+}
 
 impl NoiseHandshake {
-    pub fn listen(params: &HandshakeParams, stream: TcpStream) -> HandshakeResult {
-        listen_handshake(stream, params)
+    pub fn initiator(params: &HandshakeParams) -> Self {
+        let noise = NoiseWrapper::initiator(params);
+        NoiseHandshake {
+            noise,
+            max_message_len: params.max_message_len,
+        }
     }
 
-    pub fn send(params: &HandshakeParams, stream: TcpStream) -> HandshakeResult {
-        send_handshake(stream, params)
+    pub fn responder(params: &HandshakeParams) -> Self {
+        let noise = NoiseWrapper::responder(params);
+        NoiseHandshake {
+            noise,
+            max_message_len: params.max_message_len,
+        }
     }
-}
 
-fn listen_handshake(stream: TcpStream, params: &HandshakeParams) -> HandshakeResult {
-    let max_message_len = params.max_message_len;
-    let mut noise = NoiseWrapper::responder(params);
-    let framed = read(stream).and_then(move |(stream, msg)| {
-        let _buf = noise.read_handshake_msg(&msg);
-        write_handshake_msg(&mut noise)
+    fn read_handshake_msg<S: AsyncRead + 'static>(
+        mut self,
+        stream: S,
+    ) -> impl Future<Item = (S, Self), Error = io::Error> {
+        read(stream).and_then(move |(stream, msg)| {
+            self.noise.read_handshake_msg(&msg)?;
+            Ok((stream, self))
+        })
+    }
+
+    fn write_handshake_msg<S: AsyncWrite + 'static>(
+        mut self,
+        stream: S,
+    ) -> impl Future<Item = (S, Self), Error = io::Error> {
+        done(self.noise.write_handshake_msg())
+            .map_err(|e| e.into())
             .and_then(|(len, buf)| write(stream, &buf, len))
```

Comment:
Is it possible to rewrite `noise.write_handshake_msg()` / `noise.write` / `write` in order for them to return (or accept, in case of `write`), `Vec<u8>` instead of a length-buffer tuple? Maybe, it's my problem not being as familiar with C-style programming, but imo this code style significantly complicates understanding of what's going on.